[iOS][Secure Paste] Custom edit menu actions

jingshao-code opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

This PR implements support for custom action items in the native edit menu on iOS, with changes to both framework and engine. This PR will be updated incrementally until the full feature is complete.

Phase 1: Add hardcoded custom menu item for iOS edit menu.
Phase 2: Add Framework API for custom iOS context menu items.
Phase 3: Add Optimization, Testing, and Documentation. 

Part of https://github.com/flutter/flutter/issues/103163 

Part of https://github.com/flutter/flutter/issues/140184

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

jingshao-code committed:
[iOS] Phase 1: Add hardcoded custom menu item for iOS edit menu

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -1056,6 +1056,13 @@ - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
+- (void)flutterTextInputView:(FlutterTextInputView*)textInputView
+         performCustomAction:(NSString*)customAction
+                  withClient:(int)client {
+  [self.textInputChannel invokeMethod:@"TextInputClient.performCustomAction"
```
can you use `platformChannel` and `ContextMenu` prefix? (similar to `willDismissEditMenuWithTextInputClient` method in this file)

jingshao-code replied:
Thanks for reviewing! I've implemented all your suggestions, but I have a question about the method naming: The current method name is quite long. Would you prefer to keep it as is for clarity, or should we consider simplifying it? For example: performContextMenuCustomAction?

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -1056,6 +1056,13 @@ - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
+         performCustomAction:(NSString*)customAction
+                  withClient:(int)client {
+  [self.textInputChannel invokeMethod:@"TextInputClient.performCustomAction"
+                             arguments:@[ @(client), customAction ]];
```
is the customAction the ID? if so you should call it `customActionID`

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -996,6 +997,20 @@ - (UIMenu*)editMenuInteraction:(UIEditMenuInteraction*)interaction
+      if (customId && title) {
+        UIAction* action = [UIAction actionWithTitle:title
+                                               image:nil
+                                          identifier:customId
```
no need to set `identifier`.

hellohuanlin commented:
The engine part is heading the right direction. Good work!

jingshao-code committed:
[iOS] Address custom menu PR feedback: method naming, parameter rename, nil identifier, weak self

jingshao-code committed:
Implement Phase 2: Framework API for custom context menu items

- Add IOSSystemContextMenuItemCustom widget class
- Add IOSSystemContextMenuItemDataCustom data class
- Implement callback registration and management
- Support both API and hardcoded test callbacks

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1968,6 +1968,23 @@ class TextInput {
+  void _handleCustomAction(String actionId) {
+    // Support dynamic callbacks (new API)
+    final VoidCallback? callback = SystemContextMenuController._customActionCallbacks[actionId];
+    if (callback != null) {
+      callback();
+      return;
+    }
```
Can you also assert that the callback is not null so that developers can be aware that something is wrong?

justinmc replied:
Not sure if this is testable or not, but it would be nice if we can test it.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -2118,6 +2135,9 @@ class TextInput {
           firstArg['action'] as String,
           firstArg['data'] == null ? <String, dynamic>{} : firstArg['data'] as Map<String, dynamic>,
         );
+      case 'TextInputClient.performCustomAction':
```
Maybe this should go in services/binding.dart, where we handle onDismissSystemContextMenu.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -2624,6 +2644,9 @@ class SystemContextMenuController with SystemContextMenuClient, Diagnosticable {
   static SystemContextMenuController? _lastShown;
 
+  /// Map to store custom action callbacks, keyed by their unique identifiers.
+  static final Map<String, VoidCallback> _customActionCallbacks = <String, VoidCallback>{};
```
If the method channel call were handled in binding.dart, then you could handle this stuff via SystemContextMenuClient and SystemContextMenuController.

justinmc commented on packages/flutter/lib/src/widgets/system_context_menu.dart:
```diff
@@ -437,5 +437,63 @@ final class IOSSystemContextMenuItemLiveText extends IOSSystemContextMenuItem {
+
+  @override
+  IOSSystemContextMenuItemData getData(WidgetsLocalizations localizations) {
+    final String callbackId = '${DateTime.now().millisecondsSinceEpoch}_${title.hashCode}';
```
I think just hashCode is probably a good key for this. The only possible collision would be for the exact same item with the exact same onPressed, say if you passed the same item into the `items` List twice. Even in that case, it would be fine to have only one entry in the _customActionCallbacks map and to call the same one for both.

Also IOSSystemContextMenuItemCustom is const, so this is equivalent to passing the same instance:

```dart
items: <IOSSystemContextMenuItem>[
  IOSSystemContextMenuItemCustom(title: 'press me', onPressed: _myOnPressed),
  IOSSystemContextMenuItemCustom(title: 'press me', onPressed: _myOnPressed),
],
```

But even in that case, it's fine to have only one entry in _customActionCallbacks.

jingshao-code replied:
Looking forward to your feedback!

justinmc reviewed (commented):
Overall this approach looks solid! Most of my comments are just about how we organize the code.

I left one comment about adding an example to the examples directory. I think that will help us to visualize how this will work better, and it will allow you to remove your debug code for the vibrate button.

I also left a comment about the callback map not being cleared when the system hides the menu. I'll keep an eye out for other sources of leaks. We'll have to make sure that we cover use cases like showing one SystemContextMenu widget, then immediately showing another. I think `hide` might get called in that case so we'll be covered, but we should make sure.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -2809,6 +2853,7 @@ class SystemContextMenuController with SystemContextMenuClient, Diagnosticable {
     }
     _lastShown = null;
     ServicesBinding.systemContextMenuClient = null;
+    _clearCustomActions();
```
This is how the platform hides it: https://github.com/flutter/flutter/blob/323b45e9949d0b1fccce9137895a72aed07efe93/packages/flutter/lib/src/services/binding.dart#L415

jingshao-code committed:
Address Justin feedback: refactor custom menu implementation

- Instance-based callback management in SystemContextMenuController

- Route custom actions through platform channel via binding.dart

- Clear callbacks on hide() and handleSystemHide()

- Remove getData() side effects, use hashCode for IDs

- Add example (system_context_menu.1.dart) and tests

jingshao-code committed:
resolve conflicts

justinmc commented on examples/api/lib/widgets/system_context_menu/system_context_menu.1.dart:
This is a great example, thanks for adding it.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -2770,9 +2783,18 @@ class SystemContextMenuController with SystemContextMenuClient, Diagnosticable {
 
     ServicesBinding.systemContextMenuClient = this;
 
+    _customActionCallbacks.clear();
```
I think you also might need to call this in handleSystemHide. Then I think we will be covered and this shouldn't leak.

justinmc replied:
Also test this: Show the SystemContextMenu with a certain item, then show again with a new item. Then try sending a handleCustomContextMenuAction for the original item.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -2770,9 +2783,18 @@ class SystemContextMenuController with SystemContextMenuClient, Diagnosticable {
+    
+    for (final IOSSystemContextMenuItemData item in items) {
+      if (item is IOSSystemContextMenuItemDataCustom) {
+        _customActionCallbacks[item.callbackId] = item.onPressed;
```
What do you think, would it be worth it to assert that there are no duplicate `item.callbackIds`? I guess it's not a problem for us if there are, and maybe users have a real reason to do something like that.

justinmc replied:
Maybe assert if there are duplicate callbackIds AND the callbacks are not equivalent? That seems like it would very likely be a mistake on the part of the app developer.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -3081,5 +3104,66 @@ final class IOSSystemContextMenuItemDataLiveText extends IOSSystemContextMenuIte
+  @override
+  Map<String, dynamic> get _json {
+    return <String, dynamic>{
+      'callbackId': callbackId,
+      'title': title,
+      'type': _jsonType,
+      'id': callbackId,
```
Do you need both an id and a callbackId?

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -3081,5 +3104,66 @@ final class IOSSystemContextMenuItemDataLiveText extends IOSSystemContextMenuIte
+    if (identical(this, other)) {
+      return true;
+    }
+    return other is IOSSystemContextMenuItemDataCustom &&
+        other.title == title &&
+        other.callbackId == callbackId;
```
onPressed should probably also be included here.

justinmc replied:
Also should include onPressed in hashCode (discussed offline).

justinmc commented on packages/flutter/lib/src/widgets/system_context_menu.dart:
```diff
@@ -458,5 +458,65 @@ final class IOSSystemContextMenuItemLiveText extends IOSSystemContextMenuItem {
+  IOSSystemContextMenuItemData getData(WidgetsLocalizations localizations) {
+    return IOSSystemContextMenuItemDataCustom(
+      title: title,
+      callbackId: hashCode.toString(),
```
Instead of passing the callbackId in here, could IOSSystemContextMenuItemData generate it itself?

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -895,4 +895,121 @@ void main() {
+      final IOSSystemContextMenuItemCustom customItem1 = items[1] as IOSSystemContextMenuItemCustom;
+      final IOSSystemContextMenuItemCustom customItem2 = items[2] as IOSSystemContextMenuItemCustom;
+      
+      controller2.handleCustomContextMenuAction(customItem1.hashCode.toString());
```
Can you also test that a call from onPerformCustomAction works? You can simulate the platform message using [TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger.handlePlatformMessage](https://main-api.flutter.dev/flutter/flutter_test/TestDefaultBinaryMessenger/handlePlatformMessage.html).

No need to use handlePlatformMessage every time, it's also fine to use handleCustomContextMenuAction directly like you are here. Just make sure that the onPerformCustomAction gets tested at least once too.

justinmc replied:
Also try the error state: send that message when there is no _systemContextMenuClient.

justinmc commented on packages/flutter/lib/src/widgets/system_context_menu.dart:
```diff
@@ -458,5 +458,65 @@ final class IOSSystemContextMenuItemLiveText extends IOSSystemContextMenuItem {
+///
+/// The [title] and [onPressed] callback must be provided.
+///
+/// {@tool dartpad}
+/// This example shows how to add custom menu items to the iOS system context menu.
+///
+/// ** See code in examples/api/lib/widgets/system_context_menu/system_context_menu.1.dart **
+/// {@end-tool}
```
Do you think it would be valuable to also add this example to SystemContextMenu.items, or is it too much?

justinmc reviewed (commented):
This is looking good, thanks for doing that refactor. I think putting the callback map inside of SystemContextMenuController is much better now. Also the example is great.

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -895,4 +895,121 @@ void main() {
+      expect(customAction1Called, isTrue);
+      expect(customAction2Called, isFalse);
+      
+      controller2.handleCustomContextMenuAction(customItem2.hashCode.toString());
```
Probably put those tests in system_context_menu_controller_test.dart.

jingshao-code committed:
Address PR feedback: refactor custom menu implementation

- Use atomic text editing value updates
- Make callback ID generation internal to IOSSystemContextMenuItemDataCustom
- Simplify JSON output to use single ID field
- Fix hashCode and equality to include onPressed callback
- Add assertion for duplicate callback IDs
- Update tests to use TargetPlatformVariant
- Add platform channel integration test
- Add edge case tests for error scenarios
- Add documentation example for custom menu items

justinmc commented on packages/flutter/test/services/system_context_menu_controller_test.dart:
```diff
@@ -546,4 +546,310 @@ void main() {
+      controller.dispose();
+    });
+    
+    // First menu with old action
```
Nit: Some of these comments are missing a period at the end.

justinmc commented on packages/flutter/test/services/system_context_menu_controller_test.dart:
Thanks for testing all of these different cases here! I think you got everything I can think of.

justinmc commented on examples/api/test/widgets/system_context_menu/system_context_menu.1_test.dart:
```diff
@@ -0,0 +1,120 @@
+      await tester.pumpAndSettle();
+
+      expect(controller.text, '');
+      expect(find.text('Text cleared'), findsOneWidget);
```
Can you also check whether the SystemContextMenu is still visible or not? I assume the answer is no, it hides itself? If not the we should show how to do that in the example, but I think it does hide.

justinmc reviewed (commented):
I had a question about whether the menu auto-closes when a button is pressed or not that came up in a few comments below.

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -895,4 +895,203 @@ void main() {
+    variant: TargetPlatformVariant.only(TargetPlatform.iOS),
+  );
+
+  testWidgets(
+    'can trigger custom menu action through platform channel message',
```
If not you could consider deleting it, but it's up to you if this is valuable or not.

jingshao-code committed:
Fix naming consistency, add tests, and improve documentation

- Verify iOS system menu auto-closes after custom action via device testing
- Change actionId to callbackId throughout for consistency
- Add test for menu auto-close behavior after custom action
- Add test for two TextFields with different custom menus
- Fix documentation to say 'title and onPressed' instead of 'title and callbackId'
- Add AdaptiveTextSelectionToolbar check in example test
- Improve error message in ServicesBinding with actionable advice
- Add missing periods to comments

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -3081,5 +3109,65 @@ final class IOSSystemContextMenuItemDataLiveText extends IOSSystemContextMenuIte
+  final VoidCallback onPressed;
+
+  /// The unique identifier for this custom action.
+  String get callbackId => hashCode.toString();
```
I'm on board with that 👍

justinmc replied:
Nit: Did you think about making callbackId an int instead of a String? I think a String makes sense to me, just wanted to make sure you thought through it.

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -895,4 +895,400 @@ void main() {
+      
+      expect(customActionCalled, isTrue);
+      
+      // iOS system menus auto-close after custom actions on real devices,
+      // but in tests we need to manually call hideToolbar() to simulate this.
+      state.hideToolbar();
```
Instead of calling hideToolbar would it be possible to send the `ContextMenu.onDismissSystemContextMenu` platform channel message in order to make this more realistic?

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -895,4 +895,400 @@ void main() {
+      expect(field2ActionCalled, isTrue);
+      
+      state2.hideToolbar();
+      await tester.pump();
```
Nit: Maybe do a final expect that SystemContextMenu is not found. Ending on a `pump` makes me wonder what was being tested there.

justinmc reviewed (approved):
LGTM with nits 👍 . Great work!

justinmc commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1003,6 +1004,24 @@ - (UIMenu*)editMenuInteraction:(UIEditMenuInteraction*)interaction
                               suggestedMenu:suggestedMenu];
       }
+    } else if ([type isEqualToString:@"custom"]) {
+      NSString* customId = encodedItem[@"id"];
```
Nit: Have you thought about naming this parameter callbackId just so it has the same name as it does in the framework? I'm not sure which name is better, just wanted to call it out.

jingshao-code committed:
- Replace hideToolbar() with onDismissSystemContextMenu platform message in tests
- Add final expect verification in two TextFields test
- Rename customId to callbackId in engine for consistency

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -1060,6 +1060,13 @@ - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
 }
 
+- (void)flutterTextInputView:(FlutterTextInputView*)textInputView
+    performPlatformChannelContextMenuCustomAction:(NSString*)customActionID
```
nit: can you call it `performContextMenuCustomActionWithActionID:(NSString*)customActionID textInputClient:(int)client`?

hellohuanlin reviewed (commented):
@jingshao-code nice work. can you add some unit tests on the native side?

jingshao-code committed:
Rename delegate method and add native tests

- Rename delegate method to performContextMenuCustomActionWithActionID:textInputClient:
- Add 2 native-side unit tests for custom menu functionality

jingshao-code committed:
Fix framework-side issues for CI. No functional changes, only code style fixes to pass CI checks

jingshao-code committed:
Fix formatting issues in framework-side

jingshao-code committed:
Fix method name and arguments in context menu tests

jingshao-code committed:
Fix test codec and multiple controllers logic

jingshao-code committed:
Fix test: Add platform dismiss message for iOS menu auto-close

jingshao-code committed:
Retry CI

hellohuanlin commented:
@jingshao-code how did you do the rebase? Try squashing your commits into 1, then rebase (not merge) on top of the main branch. You can also cherry-pick after squash.

jingshao-code committed:
Fix CI test failures in system context menu

Add custom type support to systemContextMenuItemDataFromJson with fake callback

hellohuanlin reviewed (approved):
LGTM!

jingshao-code committed:
Fix formatting and custom menu action test

hellohuanlin commented:
> The error is about bad json:
> 
> Error
> The cause seems to be that systemContextMenuItemDataFromJson doesn't have support for the custom type:

Oh I missed it when reading the failure. Thanks for spotting it!

jingshao-code committed:
Fix custom menu test formatting and remove redundant mock

jingshao-code committed:
Fix custom menu test mock setup

jingshao-code committed:
Fix OCMock andDo macro compilation error

justinmc reviewed (approved):
LGTM with nits 👍 

I also had a comment about renaming "id" to "callbackId", did you intend to do that or decide to leave it? https://github.com/flutter/flutter/pull/171825#discussion_r2267950975

jingshao-code committed:
test

jingshao-code committed:
Trigger Build

jingshao-code committed:
Merge branch 'master' into ios-custom-menu
